cli/core/test/e2e/framework/plugin_operations.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package framework
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	plugintypes "github.com/vmware-tanzu/tanzu-framework/apis/cli/v1alpha1"
+)
+
+// PluginOps helps to perform plugin operations like create plugin bundle, and publish
+type PluginOps interface {
+
+	// GenerateScriptBasedPluginBinaries generates script based plugin binaries for given plugin metadata
+	GenerateScriptBasedPluginBinaries(pluginsMeta []*PluginMeta) ([]string, []error)
+
+	// PublishPluginBinary publishes the plugin binaries to given registry bucket and returns the plugin distribution urls
+	PublishPluginBinary(pluginsInfo []*PluginMeta) (discoveryUrls []string, errs []error)
+
+	// GeneratePluginBundle generates plugin bundle in local file system for given plugin metadata
+	GeneratePluginBundle(pluginsMeta []*PluginMeta) ([]string, []error)
+
+	// PublishPluginBundle publishes the plugin bundles to given registry bucket and returns the plugins discovery urls
+	PublishPluginBundle(pluginsInfo []*PluginMeta) (discoveryUrls []string, errs []error)
+}
+
+// CLIPlugin for plugin overlay info
+type CLIPlugin struct {
+	metav1.TypeMeta `yaml:",inline"`
+	Metadata        metav1.ObjectMeta         `json:"metadata"`
+	Spec            plugintypes.CLIPluginSpec `json:"spec"`
+}
+
+// localOCIPluginOps is the implementation of PluginOps interface
+type localOCIPluginOps struct {
+	PluginOps
+	cmdExe    CmdOps
+	registry  PluginRegistry
+	imgpkgOps ImgpkgOps
+}
+
+func NewLocalOCIPluginOps(registry PluginRegistry) PluginOps {
+	return &localOCIPluginOps{
+		cmdExe:    NewCmdOps(),
+		registry:  registry,
+		imgpkgOps: NewImgpkgOps(),
+	}
+}
+
+func (po *localOCIPluginOps) GenerateScriptBasedPluginBinaries(pluginsMeta []*PluginMeta) ([]string, []error) {
+	pluginsProcessed := make(map[string]bool)
+	size := len(pluginsMeta)
+	pluginBinaryFilePaths := make([]string, size)
+	errs := make([]error, size)
+	for i, pm := range pluginsMeta {
+		if _, exists := pluginsProcessed[pm.name]; exists {
+			errs[i] = errors.New("plugin name already exists, currently multiple versions of same plugin not supported")
```

Comment:
The uniqueness of a plugin is based on the plugin name and target. It is possible that I want to create 2 plugin binaries with the same plugin name but associated with different targets.

`PluginMeta` should include the `target` field as well. And changes need to be done accordingly in these functions as well.

Reply:
changes has done to add `target` in `PluginMeta`